nni/mutable/mutable.py

```diff
@@ -501,3 +508,608 @@ class MutableSymbol(LabeledMutable, Symbol, MutableExpression):
+        label: str | None = None
+    ) -> None:
+        values = list(values)
+        assert values, 'Discrete values must not be empty.'
+        self.label = label or auto_label()
+        self.values = values
+        self.distribution = distribution if distribution is not None else [1 / len(values)] * len(values)
+
+        if default is not MISSING:
+            self.validate({self.label: default})
+        self.default_value = default
+
+        assert not(any(isinstance(value, Mutable) for value in values)), 'Discrete values must not contain mutables.'
+        assert len(set(values)) == len(values), 'Values must be unique.'
+        assert len(self.distribution) == len(self.values), 'Distribution must have length n.'
+        assert abs(sum(self.distribution) - 1) < 1e-6, 'Distribution must sum to 1.'
+
+    def contains(self, sample: Sample) -> SampleValidationError | None:
+        if self.label not in sample:
+            return SampleMissingError(self.label, list(sample.keys()))
+        sample_val = sample[self.label]
+        if sample_val not in self.values:
+            return SampleValidationError(f'{sample_val} not found in {self.values}')
+        return None
+
+    def extra_repr(self) -> str:
+        if len(self.values) <= 7:
+            return f'{self.values!r}, label={self.label!r}'
+        return '[' +  \
+            ', '.join(map(repr, self.values[:3])) + \
+            ', ..., ' + \
+            ', '.join(map(repr, self.values[-3:])) + \
+            f'], label={self.label!r}'
+
+    def freeze(self, sample: Sample) -> Any:
+        self.validate(sample)
+        return sample[self.label]
+
+    def __len__(self):
+        return len(self.values)
+
+    def default(self, memo: Sample | None = None) -> Choice:
+        """The default() of :class:`Discrete` is the first value unless default value is set.
+
+        See Also
+        --------
+        Mutable.default
+        """
+        memo = {} if memo is None else memo
+        err = self.contains(memo)
+        if isinstance(err, SampleMissingError):
+            if self.default_value is not MISSING:
+                memo[self.label] = self.default_value
+            else:
+                memo[self.label] = self.values[0]
+        rv = self.freeze(memo)
+        if self.default_value is not MISSING and rv != self.default_value:
+            raise ValueError(f'Default value is specified to be {self.default_value} but got {rv}. '
+                             f'Please check the default value of {self.label}.')
+        return rv
+
+    def random(self, memo: Sample | None = None, random_state: RandomState | None = None) -> Choice:
+        """Randomly sample a value from choices.
+        Distribution is respected if provided.
+
+        See Also
+        --------
+        Mutable.random
+        """
+        memo = {} if memo is None else memo
+        if random_state is None:
+            random_state = RandomState()
+        err = self.contains(memo)
```

Comment:
I just found that `if self.contains(sample):` evaluates to True when it does not contain sample...

Reply:
`check_contains()` should be better.
I will check the doc when I see a strange name like this.